# Part C: Text Classification with RNNs

This notebook ports `rnn.py` into a structured notebook and extends it with multiple RNN/LSTM architectures,
t-SNE visualization of learned embeddings, and several experimental variations.

## 1. Imports & Configuration

In [2]:
%pip install torch==2.3.0 torchtext==0.18.0 --index-url https://download.pytorch.org/whl/cpu --force-reinstall
%pip install pandas numpy tqdm scikit-learn matplotlib seaborn gensim

Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://download.pytorch.org/whl/cpu
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 190.4/190.4 MB 11.1 MB/s eta 0:00:0000:0100:01
  Using cached https://download.pytorch.org/whl/cpu/torchtext-0.18.0%2Bcpu-cp310-cp310-linux_x86_64.whl (2.0 MB)
  Using cached filelock-3.25.2-py3-none-any.whl.metadata (2.0 kB)
  Using cached sympy-1.14.0-py3-none-any.whl.metadata (12 kB)
  Using cached networkx-3.4.2-py3-none-any.whl.metadata (6.3 kB)
  Using cached https://download.pytorch.org/whl/jinja2-3.1.6-py3-none-any.whl.metadata (2.9 kB)
  Using cached fsspec-2026.2.0-py3-none-any.whl.metadata (10 kB)
  Using cached https://download.pytorch.org/whl/tqdm-4.66.5-py3-none-any.whl (78 kB)
  Using cached https://download.pytorch.org/whl/requests-2.28.1-py3-none-any.whl (62 kB)
  Using cached numpy-2.2.6-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)
  Using cached https://downloa

In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torchtext.data import get_tokenizer
from torchtext.vocab import build_vocab_from_iterator
import pandas as pd
import numpy as np
import time
import copy
from tqdm import tqdm
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.manifold import TSNE
import matplotlib.pyplot as plt
import seaborn as sns
import gensim.downloader as api

sns.set_style('whitegrid')
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f'Using device: {device}')

/home/george/.local/lib/python3.10/site-packages/torchtext/data/__init__.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)
/home/george/.local/lib/python3.10/site-packages/torchtext/vocab/__init__.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /!\ 
Torchtext is deprecated and the last released version will be 0.18 (this one). You can silence this warning by calling the following at the beginnign of your scripts: `import torchtext; torchtext.disable_torchtext_deprecation_warning()`
  warnings.warn(torchtext._TORCHTEXT_DEPRECATION_MSG)
/home/george/.local/lib/python3.10/site-packages/torchtext/utils.py:4: UserWarning: 
/!\ IMPORTANT WARNING ABOUT TORCHTEXT STATUS /

Using device: cpu


## 2. Hyperparameters

Toggle **`MAX_WORDS`**, **`USE_GLOVE`**, **`FREEZE_EMBEDDINGS`**, and **`USE_IMDB`** to run the variations.

In [2]:
# ============================================================
# TOGGLEABLE CONFIGURATION — change these to run variations
# ============================================================
MAX_WORDS = 25            # Set to 50 for the MAX_WORDS variation
EPOCHS = 15
LEARNING_RATE = 1e-3
BATCH_SIZE = 1024
EMBEDDING_DIM = 100
HIDDEN_DIM = 64
NUM_RUNS = 3              # Number of repeated runs per architecture

USE_GLOVE = False         # Set True to init embeddings with glove-6B-100d
FREEZE_EMBEDDINGS = False # Set True to freeze pre-trained embeddings
USE_IMDB = False          # Set True to use IMDB dataset instead of AG News

print(f'MAX_WORDS={MAX_WORDS}, USE_GLOVE={USE_GLOVE}, FREEZE={FREEZE_EMBEDDINGS}, USE_IMDB={USE_IMDB}')

MAX_WORDS=25, USE_GLOVE=False, FREEZE=False, USE_IMDB=False


## 3. Data Loading

In [3]:
tokenizer = get_tokenizer('basic_english')

if USE_IMDB:
    # IMDB dataset (2 classes)
    from torch.utils.data.dataset import random_split
    imdb_data = pd.read_csv('IMDB Dataset.csv')  # expects 'review' and 'sentiment' columns
    # If file not found, try downloading:
    # imdb_data = pd.read_csv('https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz') 
    # For simplicity, we assume the CSV is available
    imdb_data['label'] = (imdb_data['sentiment'] == 'positive').astype(int)
    all_dataset = [(row['label'], row['review'].lower()) for _, row in imdb_data.iterrows()]
    train_size = int(0.8 * len(all_dataset))
    test_size = len(all_dataset) - train_size
    train_dataset, test_dataset = random_split(all_dataset, [train_size, test_size],
                                               generator=torch.Generator().manual_seed(42))
    train_dataset = list(train_dataset)
    test_dataset = list(test_dataset)
    target_classes = ['Negative', 'Positive']
    print(f'IMDB dataset loaded: {len(train_dataset)} train, {len(test_dataset)} test')
else:
    # AG News dataset (4 classes)
    train_data = pd.read_csv('ag-news-classification-dataset/train.csv')
    test_data = pd.read_csv('ag-news-classification-dataset/test.csv')
    train_dataset = [(label, train_data['Title'][i] + ' ' + train_data['Description'][i])
                     for i, label in enumerate(train_data['Class Index'])]
    test_dataset = [(label, test_data['Title'][i] + ' ' + test_data['Description'][i])
                    for i, label in enumerate(test_data['Class Index'])]
    target_classes = ['World', 'Sports', 'Business', 'Sci/Tech']
    print(f'AG News dataset loaded: {len(train_dataset)} train, {len(test_dataset)} test')

NUM_CLASSES = len(target_classes)
print(f'Classes ({NUM_CLASSES}): {target_classes}')

AG News dataset loaded: 120000 train, 7600 test
Classes (4): ['World', 'Sports', 'Business', 'Sci/Tech']


## 4. Vocabulary Building & Preprocessing

In [4]:
def build_vocabulary(datasets):
    for dataset in datasets:
        for _, text in dataset:
            yield tokenizer(text)

vocab = build_vocab_from_iterator(
    build_vocabulary([train_dataset, test_dataset]),
    min_freq=10,
    specials=['<PAD>', '<UNK>']
)
vocab.set_default_index(vocab['<UNK>'])
print(f'Vocabulary size: {len(vocab)}')

# Label offset: AG News labels are 1-indexed, IMDB are already 0-indexed
LABEL_OFFSET = 0 if USE_IMDB else 1

def collate_batch(batch):
    Y, X = list(zip(*batch))
    Y = torch.tensor(Y) - LABEL_OFFSET
    X = [vocab(tokenizer(text)) for text in X]
    X = [tokens + ([vocab['<PAD>']] * (MAX_WORDS - len(tokens))) if len(tokens) < MAX_WORDS
         else tokens[:MAX_WORDS] for tokens in X]
    return torch.tensor(X, dtype=torch.int32).to(device), Y.to(device)

train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True, collate_fn=collate_batch)
test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False, collate_fn=collate_batch)

Vocabulary size: 21254


## 5. Model Definition

A unified RNN model class supporting 6 variations via parameters:
- `rnn_type`: `'RNN'` or `'LSTM'`
- `bidirectional`: `True` / `False`
- `num_layers`: `1` or `2`

In [5]:
class RNNClassifier(nn.Module):
    def __init__(self, vocab_size, embedding_dim, hidden_dim, output_dim,
                 rnn_type='RNN', bidirectional=False, num_layers=1,
                 pretrained_embeddings=None, freeze_embeddings=False):
        super(RNNClassifier, self).__init__()
        self.embedding_layer = nn.Embedding(num_embeddings=vocab_size, embedding_dim=embedding_dim)

        # Optionally load pre-trained embeddings
        if pretrained_embeddings is not None:
            self.embedding_layer.weight.data.copy_(pretrained_embeddings)
            if freeze_embeddings:
                self.embedding_layer.weight.requires_grad = False

        rnn_cls = nn.LSTM if rnn_type == 'LSTM' else nn.RNN
        self.rnn = rnn_cls(input_size=embedding_dim, hidden_size=hidden_dim,
                           num_layers=num_layers, batch_first=True,
                           bidirectional=bidirectional)

        linear_input_dim = hidden_dim * 2 if bidirectional else hidden_dim
        self.linear = nn.Linear(linear_input_dim, output_dim)
        self.rnn_type = rnn_type

    def forward(self, X_batch):
        embeddings = self.embedding_layer(X_batch)
        output, hidden = self.rnn(embeddings)
        # Use the last time-step output for classification
        logits = self.linear(output[:, -1])
        probs = F.softmax(logits, dim=1)
        return probs

def count_parameters(model):
    return sum(p.numel() for p in model.parameters() if p.requires_grad)

print('Model class defined.')

Model class defined.


## 6. (Optional) Load Pre-trained GloVe Embeddings

In [6]:
pretrained_vectors = None

if USE_GLOVE:
    print('Loading GloVe 6B-100d...')
    glove = api.load('glove-wiki-gigaword-100')
    pretrained_vectors = torch.zeros(len(vocab), EMBEDDING_DIM)
    found = 0
    for idx, word in enumerate(vocab.get_itos()):
        if word in glove:
            pretrained_vectors[idx] = torch.tensor(glove[word], dtype=torch.float32)
            found += 1
    print(f'GloVe vectors loaded. {found}/{len(vocab)} words found.')
else:
    print('Skipping GloVe (USE_GLOVE=False)')

Skipping GloVe (USE_GLOVE=False)


## 7. Training & Evaluation Functions

In [7]:
def train_model(model, loss_fn, optimizer, train_loader, epochs):
    epoch_times = []
    for epoch in range(1, epochs + 1):
        model.train()
        losses = []
        start = time.time()
        for X, Y in tqdm(train_loader, desc=f'Epoch {epoch}', leave=False):
            preds = model(X)
            loss = loss_fn(preds, Y)
            losses.append(loss.item())
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
        elapsed = time.time() - start
        epoch_times.append(elapsed)
        print(f'Epoch {epoch:2d} | Loss: {np.mean(losses):.4f} | Time: {elapsed:.1f}s')
    return epoch_times


def evaluate_model(model, loss_fn, test_loader):
    model.eval()
    with torch.no_grad():
        Y_actual, Y_preds, losses = [], [], []
        for X, Y in test_loader:
            preds = model(X)
            loss = loss_fn(preds, Y)
            losses.append(loss.item())
            Y_actual.append(Y)
            Y_preds.append(preds.argmax(dim=-1))
        Y_actual = torch.cat(Y_actual)
        Y_preds = torch.cat(Y_preds)
    return (torch.tensor(losses).mean().item(),
            Y_actual.detach().cpu().numpy(),
            Y_preds.detach().cpu().numpy())

## 8. Define the 6 Architectures

In [8]:
architectures = [
    {'name': '1RNN',       'rnn_type': 'RNN',  'bidirectional': False, 'num_layers': 1},
    {'name': '1Bi-RNN',    'rnn_type': 'RNN',  'bidirectional': True,  'num_layers': 1},
    {'name': '2Bi-RNN',    'rnn_type': 'RNN',  'bidirectional': True,  'num_layers': 2},
    {'name': '1LSTM',      'rnn_type': 'LSTM', 'bidirectional': False, 'num_layers': 1},
    {'name': '1Bi-LSTM',   'rnn_type': 'LSTM', 'bidirectional': True,  'num_layers': 1},
    {'name': '2Bi-LSTM',   'rnn_type': 'LSTM', 'bidirectional': True,  'num_layers': 2},
]

print('Architectures to evaluate:')
for a in architectures:
    print(f"  - {a['name']}")

Architectures to evaluate:
  - 1RNN
  - 1Bi-RNN
  - 2Bi-RNN
  - 1LSTM
  - 1Bi-LSTM
  - 2Bi-LSTM


## 9. Run Evaluation Loop (3 runs × 6 architectures)

In [9]:
all_results = {}
trained_models = {}  # Store last trained model per architecture

for arch in architectures:
    arch_name = arch['name']
    print(f"\n{'='*70}")
    print(f"  Architecture: {arch_name}")
    print(f"{'='*70}")

    run_accs = []
    run_times = []
    n_params = None

    for run_idx in range(1, NUM_RUNS + 1):
        print(f"\n  --- Run {run_idx}/{NUM_RUNS} ---")

        model = RNNClassifier(
            vocab_size=len(vocab),
            embedding_dim=EMBEDDING_DIM,
            hidden_dim=HIDDEN_DIM,
            output_dim=NUM_CLASSES,
            rnn_type=arch['rnn_type'],
            bidirectional=arch['bidirectional'],
            num_layers=arch['num_layers'],
            pretrained_embeddings=pretrained_vectors,
            freeze_embeddings=FREEZE_EMBEDDINGS
        ).to(device)

        if n_params is None:
            n_params = count_parameters(model)
            print(f'  Parameters: {n_params:,}')

        loss_fn = nn.CrossEntropyLoss()
        optimizer = torch.optim.Adam(
            [p for p in model.parameters() if p.requires_grad],
            lr=LEARNING_RATE
        )

        epoch_times = train_model(model, loss_fn, optimizer, train_loader, EPOCHS)
        _, Y_actual, Y_preds = evaluate_model(model, loss_fn, test_loader)
        acc = accuracy_score(Y_actual, Y_preds)

        run_accs.append(acc)
        run_times.append(np.mean(epoch_times))
        print(f'  Run {run_idx} Test Accuracy: {acc:.4f}')

        trained_models[arch_name] = model  # keep last run

    all_results[arch_name] = {
        'mean_acc': np.mean(run_accs),
        'std_acc': np.std(run_accs),
        'params': n_params,
        'mean_time_per_epoch': np.mean(run_times),
    }

print('\n\nAll runs complete!')


  Architecture: 1RNN

  --- Run 1/3 ---
  Parameters: 2,136,284


Epoch  1 | Loss: 1.2982 | Time: 9.2s


Epoch  2 | Loss: 1.0589 | Time: 8.8s


Epoch  3 | Loss: 0.9668 | Time: 9.4s


Epoch  4 | Loss: 0.9259 | Time: 9.0s


Epoch  5 | Loss: 0.9034 | Time: 10.0s


Epoch  6 | Loss: 0.8864 | Time: 9.7s


Epoch  7 | Loss: 0.8757 | Time: 9.7s


Epoch  8 | Loss: 0.8662 | Time: 10.0s


Epoch  9 | Loss: 0.8592 | Time: 10.1s


Epoch 10 | Loss: 0.8528 | Time: 9.7s


Epoch 11 | Loss: 0.8482 | Time: 10.2s


Epoch 12 | Loss: 0.8465 | Time: 10.1s


Epoch 13 | Loss: 0.8410 | Time: 10.0s


Epoch 14 | Loss: 0.8373 | Time: 9.8s


Epoch 15 | Loss: 0.8347 | Time: 9.9s
  Run 1 Test Accuracy: 0.8630

  --- Run 2/3 ---


Epoch  1 | Loss: 1.2986 | Time: 10.1s


Epoch  2 | Loss: 1.0701 | Time: 9.8s


Epoch  3 | Loss: 0.9727 | Time: 10.0s


Epoch  4 | Loss: 0.9291 | Time: 9.6s


Epoch  5 | Loss: 0.9050 | Time: 10.1s


Epoch  6 | Loss: 0.8877 | Time: 9.7s


Epoch  7 | Loss: 0.8770 | Time: 9.8s


Epoch  8 | Loss: 0.8675 | Time: 12.7s


Epoch  9 | Loss: 0.8595 | Time: 9.7s


Epoch 10 | Loss: 0.8530 | Time: 10.3s


Epoch 11 | Loss: 0.8491 | Time: 10.1s


Epoch 12 | Loss: 0.8442 | Time: 9.5s


Epoch 13 | Loss: 0.8403 | Time: 9.2s


Epoch 14 | Loss: 0.8372 | Time: 9.3s


Epoch 15 | Loss: 0.8355 | Time: 9.5s
  Run 2 Test Accuracy: 0.8654

  --- Run 3/3 ---


Epoch  1 | Loss: 1.2992 | Time: 9.7s


Epoch  2 | Loss: 1.0481 | Time: 9.5s


Epoch  3 | Loss: 0.9609 | Time: 10.1s


Epoch  4 | Loss: 0.9210 | Time: 10.0s


Epoch  5 | Loss: 0.8981 | Time: 9.6s


Epoch  6 | Loss: 0.8822 | Time: 9.6s


Epoch  7 | Loss: 0.8709 | Time: 9.8s


Epoch  8 | Loss: 0.8622 | Time: 9.5s


Epoch  9 | Loss: 0.8546 | Time: 9.6s


Epoch 10 | Loss: 0.8480 | Time: 8.9s


Epoch 11 | Loss: 0.8456 | Time: 9.6s


Epoch 12 | Loss: 0.8401 | Time: 10.5s


Epoch 13 | Loss: 0.8355 | Time: 10.0s


Epoch 14 | Loss: 0.8331 | Time: 10.5s


Epoch 15 | Loss: 0.8298 | Time: 10.6s
  Run 3 Test Accuracy: 0.8776

  Architecture: 1Bi-RNN

  --- Run 1/3 ---
  Parameters: 2,147,164


Epoch  1 | Loss: 1.2998 | Time: 15.7s


Epoch  2 | Loss: 1.1043 | Time: 15.5s


Epoch  3 | Loss: 1.0045 | Time: 15.1s


Epoch  4 | Loss: 0.9348 | Time: 15.6s


Epoch  5 | Loss: 0.9046 | Time: 15.3s


Epoch  6 | Loss: 0.8863 | Time: 15.3s


Epoch  7 | Loss: 0.8746 | Time: 14.9s


Epoch  8 | Loss: 0.8646 | Time: 14.5s


Epoch  9 | Loss: 0.8565 | Time: 14.7s


Epoch 10 | Loss: 0.8510 | Time: 15.8s


Epoch 11 | Loss: 0.8451 | Time: 14.9s


Epoch 12 | Loss: 0.8413 | Time: 15.0s


Epoch 13 | Loss: 0.8385 | Time: 15.2s


Epoch 14 | Loss: 0.8345 | Time: 15.5s


Epoch 15 | Loss: 0.8310 | Time: 14.1s
  Run 1 Test Accuracy: 0.8705

  --- Run 2/3 ---


Epoch  1 | Loss: 1.3033 | Time: 13.3s


Epoch  2 | Loss: 1.1046 | Time: 13.1s


Epoch  3 | Loss: 0.9844 | Time: 13.1s


Epoch  4 | Loss: 0.9319 | Time: 13.2s


Epoch  5 | Loss: 0.9059 | Time: 15.7s


Epoch  6 | Loss: 0.8896 | Time: 16.5s


Epoch  7 | Loss: 0.8773 | Time: 16.7s


Epoch  8 | Loss: 0.8684 | Time: 17.4s


Epoch  9 | Loss: 0.8603 | Time: 16.4s


Epoch 10 | Loss: 0.8556 | Time: 16.7s


Epoch 11 | Loss: 0.8514 | Time: 16.3s


Epoch 12 | Loss: 0.8461 | Time: 16.5s


Epoch 13 | Loss: 0.8431 | Time: 16.6s


Epoch 14 | Loss: 0.8402 | Time: 16.6s


Epoch 15 | Loss: 0.8361 | Time: 16.6s
  Run 2 Test Accuracy: 0.8633

  --- Run 3/3 ---


Epoch  1 | Loss: 1.3028 | Time: 15.6s


Epoch  2 | Loss: 1.0728 | Time: 15.0s


Epoch  3 | Loss: 0.9694 | Time: 15.1s


Epoch  4 | Loss: 0.9269 | Time: 14.1s


Epoch  5 | Loss: 0.9041 | Time: 14.1s


Epoch  6 | Loss: 0.8872 | Time: 14.7s


Epoch  7 | Loss: 0.8757 | Time: 14.6s


Epoch  8 | Loss: 0.8675 | Time: 14.1s


Epoch  9 | Loss: 0.8588 | Time: 14.2s


Epoch 10 | Loss: 0.8525 | Time: 13.6s


Epoch 11 | Loss: 0.8492 | Time: 14.1s


Epoch 12 | Loss: 0.8448 | Time: 15.6s


Epoch 13 | Loss: 0.8411 | Time: 15.3s


Epoch 14 | Loss: 0.8377 | Time: 15.8s


Epoch 15 | Loss: 0.8342 | Time: 15.5s
  Run 3 Test Accuracy: 0.8687

  Architecture: 2Bi-RNN

  --- Run 1/3 ---
  Parameters: 2,171,996


Epoch  1 | Loss: 1.2480 | Time: 22.4s


Epoch  2 | Loss: 1.0218 | Time: 21.8s


Epoch  3 | Loss: 0.9592 | Time: 22.0s


Epoch  4 | Loss: 0.9281 | Time: 22.5s


Epoch  5 | Loss: 0.9068 | Time: 22.1s


Epoch  6 | Loss: 0.8925 | Time: 22.1s


Epoch  7 | Loss: 0.8845 | Time: 22.3s


Epoch  8 | Loss: 0.8737 | Time: 22.2s


Epoch  9 | Loss: 0.8735 | Time: 22.1s


Epoch 10 | Loss: 0.8650 | Time: 21.8s


Epoch 11 | Loss: 0.8622 | Time: 22.0s


Epoch 12 | Loss: 0.8549 | Time: 22.3s


Epoch 13 | Loss: 0.8570 | Time: 22.2s


Epoch 14 | Loss: 0.8495 | Time: 21.8s


Epoch 15 | Loss: 0.8468 | Time: 22.0s
  Run 1 Test Accuracy: 0.8612

  --- Run 2/3 ---


Epoch  1 | Loss: 1.2649 | Time: 22.2s


Epoch  2 | Loss: 1.0561 | Time: 22.3s


Epoch  3 | Loss: 0.9924 | Time: 22.1s


Epoch  4 | Loss: 0.9666 | Time: 22.3s


Epoch  5 | Loss: 0.9456 | Time: 22.1s


Epoch  6 | Loss: 0.9304 | Time: 22.1s


Epoch  7 | Loss: 0.9203 | Time: 21.8s


Epoch  8 | Loss: 0.9084 | Time: 22.3s


Epoch  9 | Loss: 0.9014 | Time: 22.0s


Epoch 10 | Loss: 0.8969 | Time: 22.0s


Epoch 11 | Loss: 0.8872 | Time: 21.9s


Epoch 12 | Loss: 0.8774 | Time: 21.9s


Epoch 13 | Loss: 0.8778 | Time: 22.2s


Epoch 14 | Loss: 0.8705 | Time: 21.8s


Epoch 15 | Loss: 0.8658 | Time: 22.0s
  Run 2 Test Accuracy: 0.8489

  --- Run 3/3 ---


Epoch  1 | Loss: 1.2594 | Time: 22.5s


Epoch  2 | Loss: 1.0484 | Time: 23.3s


Epoch  3 | Loss: 0.9758 | Time: 23.0s


Epoch  4 | Loss: 0.9348 | Time: 22.4s


Epoch  5 | Loss: 0.9137 | Time: 23.7s


Epoch  6 | Loss: 0.8948 | Time: 23.1s


Epoch  7 | Loss: 0.8861 | Time: 23.4s


Epoch  8 | Loss: 0.8782 | Time: 22.1s


Epoch  9 | Loss: 0.8748 | Time: 25.1s


Epoch 10 | Loss: 0.8772 | Time: 27.5s


Epoch 11 | Loss: 0.8650 | Time: 23.1s


Epoch 12 | Loss: 0.8571 | Time: 23.0s


Epoch 13 | Loss: 0.8719 | Time: 22.7s


Epoch 14 | Loss: 0.8633 | Time: 22.7s


Epoch 15 | Loss: 0.8559 | Time: 23.1s
  Run 3 Test Accuracy: 0.8499

  Architecture: 1LSTM

  --- Run 1/3 ---
  Parameters: 2,168,156


Epoch  1 | Loss: 1.2469 | Time: 14.9s


Epoch  2 | Loss: 0.9716 | Time: 14.8s


Epoch  3 | Loss: 0.9082 | Time: 14.7s


Epoch  4 | Loss: 0.8808 | Time: 15.2s


Epoch  5 | Loss: 0.8641 | Time: 15.2s


Epoch  6 | Loss: 0.8521 | Time: 14.9s


Epoch  7 | Loss: 0.8436 | Time: 14.8s


Epoch  8 | Loss: 0.8372 | Time: 15.1s


Epoch  9 | Loss: 0.8308 | Time: 14.9s


Epoch 10 | Loss: 0.8268 | Time: 15.1s


Epoch 11 | Loss: 0.8220 | Time: 15.1s


Epoch 12 | Loss: 0.8191 | Time: 14.8s


Epoch 13 | Loss: 0.8167 | Time: 15.0s


Epoch 14 | Loss: 0.8145 | Time: 15.3s


Epoch 15 | Loss: 0.8117 | Time: 14.9s
  Run 1 Test Accuracy: 0.8789

  --- Run 2/3 ---


Epoch  1 | Loss: 1.2520 | Time: 15.0s


Epoch  2 | Loss: 0.9773 | Time: 15.0s


Epoch  3 | Loss: 0.9093 | Time: 15.4s


Epoch  4 | Loss: 0.8814 | Time: 14.8s


Epoch  5 | Loss: 0.8632 | Time: 14.7s


Epoch  6 | Loss: 0.8514 | Time: 14.4s


Epoch  7 | Loss: 0.8420 | Time: 14.7s


Epoch  8 | Loss: 0.8359 | Time: 14.8s


Epoch  9 | Loss: 0.8299 | Time: 14.8s


Epoch 10 | Loss: 0.8256 | Time: 15.1s


Epoch 11 | Loss: 0.8211 | Time: 15.2s


Epoch 12 | Loss: 0.8175 | Time: 15.3s


Epoch 13 | Loss: 0.8154 | Time: 14.7s


Epoch 14 | Loss: 0.8132 | Time: 13.6s


Epoch 15 | Loss: 0.8109 | Time: 14.5s
  Run 2 Test Accuracy: 0.8814

  --- Run 3/3 ---


Epoch  1 | Loss: 1.2640 | Time: 14.6s


Epoch  2 | Loss: 0.9816 | Time: 15.2s


Epoch  3 | Loss: 0.9078 | Time: 15.2s


Epoch  4 | Loss: 0.8791 | Time: 14.2s


Epoch  5 | Loss: 0.8608 | Time: 14.1s


Epoch  6 | Loss: 0.8491 | Time: 13.7s


Epoch  7 | Loss: 0.8403 | Time: 14.4s


Epoch  8 | Loss: 0.8331 | Time: 13.9s


Epoch  9 | Loss: 0.8262 | Time: 13.7s


Epoch 10 | Loss: 0.8221 | Time: 13.7s


Epoch 11 | Loss: 0.8188 | Time: 13.4s


Epoch 12 | Loss: 0.8149 | Time: 13.5s


Epoch 13 | Loss: 0.8117 | Time: 13.5s


Epoch 14 | Loss: 0.8106 | Time: 13.4s


Epoch 15 | Loss: 0.8080 | Time: 13.8s
  Run 3 Test Accuracy: 0.8845

  Architecture: 1Bi-LSTM

  --- Run 1/3 ---
  Parameters: 2,210,908


Epoch  1 | Loss: 1.2533 | Time: 24.2s


Epoch  2 | Loss: 0.9792 | Time: 23.3s


Epoch  3 | Loss: 0.9111 | Time: 23.7s


Epoch  4 | Loss: 0.8818 | Time: 23.3s


Epoch  5 | Loss: 0.8644 | Time: 24.0s


Epoch  6 | Loss: 0.8524 | Time: 23.1s


Epoch  7 | Loss: 0.8430 | Time: 22.8s


Epoch  8 | Loss: 0.8352 | Time: 22.2s


Epoch  9 | Loss: 0.8296 | Time: 22.0s


Epoch 10 | Loss: 0.8247 | Time: 21.8s


Epoch 11 | Loss: 0.8203 | Time: 21.8s


Epoch 12 | Loss: 0.8177 | Time: 21.7s


Epoch 13 | Loss: 0.8147 | Time: 22.8s


Epoch 14 | Loss: 0.8121 | Time: 23.6s


Epoch 15 | Loss: 0.8102 | Time: 24.4s
  Run 1 Test Accuracy: 0.8793

  --- Run 2/3 ---


Epoch  1 | Loss: 1.2625 | Time: 27.1s


Epoch  2 | Loss: 0.9854 | Time: 26.8s


Epoch  3 | Loss: 0.9131 | Time: 25.4s


Epoch  4 | Loss: 0.8840 | Time: 25.5s


Epoch  5 | Loss: 0.8652 | Time: 26.0s


Epoch  6 | Loss: 0.8534 | Time: 25.0s


Epoch  7 | Loss: 0.8442 | Time: 23.2s


Epoch  8 | Loss: 0.8368 | Time: 23.7s


Epoch  9 | Loss: 0.8307 | Time: 26.1s


Epoch 10 | Loss: 0.8249 | Time: 24.7s


Epoch 11 | Loss: 0.8215 | Time: 23.9s


Epoch 12 | Loss: 0.8184 | Time: 24.3s


Epoch 13 | Loss: 0.8149 | Time: 23.8s


Epoch 14 | Loss: 0.8129 | Time: 22.8s


Epoch 15 | Loss: 0.8100 | Time: 22.9s
  Run 2 Test Accuracy: 0.8817

  --- Run 3/3 ---


Epoch  1 | Loss: 1.2583 | Time: 23.5s


Epoch  2 | Loss: 0.9866 | Time: 22.4s


Epoch  3 | Loss: 0.9124 | Time: 22.9s


Epoch  4 | Loss: 0.8810 | Time: 23.6s


Epoch  5 | Loss: 0.8635 | Time: 23.6s


Epoch  6 | Loss: 0.8502 | Time: 23.9s


Epoch  7 | Loss: 0.8414 | Time: 23.9s


Epoch  8 | Loss: 0.8340 | Time: 23.2s


Epoch  9 | Loss: 0.8288 | Time: 23.4s


Epoch 10 | Loss: 0.8244 | Time: 23.1s


Epoch 11 | Loss: 0.8209 | Time: 22.6s


Epoch 12 | Loss: 0.8163 | Time: 22.0s


Epoch 13 | Loss: 0.8143 | Time: 22.5s


Epoch 14 | Loss: 0.8117 | Time: 23.4s


Epoch 15 | Loss: 0.8101 | Time: 23.6s
  Run 3 Test Accuracy: 0.8803

  Architecture: 2Bi-LSTM

  --- Run 1/3 ---
  Parameters: 2,310,236


Epoch  1 | Loss: 1.1856 | Time: 39.1s


Epoch  2 | Loss: 0.9476 | Time: 41.9s


Epoch  3 | Loss: 0.8948 | Time: 41.3s


Epoch  4 | Loss: 0.8704 | Time: 42.8s


Epoch  5 | Loss: 0.8546 | Time: 42.5s


Epoch  6 | Loss: 0.8455 | Time: 42.6s


Epoch  7 | Loss: 0.8389 | Time: 43.2s


Epoch  8 | Loss: 0.8333 | Time: 44.4s


Epoch  9 | Loss: 0.8284 | Time: 43.2s


Epoch 10 | Loss: 0.8242 | Time: 41.8s


Epoch 11 | Loss: 0.8203 | Time: 42.4s


Epoch 12 | Loss: 0.8161 | Time: 44.3s


Epoch 13 | Loss: 0.8170 | Time: 47.7s


Epoch 14 | Loss: 0.8123 | Time: 38.8s


Epoch 15 | Loss: 0.8102 | Time: 40.6s
  Run 1 Test Accuracy: 0.8857

  --- Run 2/3 ---


Epoch  1 | Loss: 1.1788 | Time: 45.5s


Epoch  2 | Loss: 0.9454 | Time: 44.1s


Epoch  3 | Loss: 0.8942 | Time: 45.3s


Epoch  4 | Loss: 0.8713 | Time: 42.4s


Epoch  5 | Loss: 0.8553 | Time: 43.4s


Epoch  6 | Loss: 0.8450 | Time: 44.2s


Epoch  7 | Loss: 0.8367 | Time: 43.4s


Epoch  8 | Loss: 0.8308 | Time: 44.6s


Epoch  9 | Loss: 0.8254 | Time: 42.3s


Epoch 10 | Loss: 0.8204 | Time: 43.0s


Epoch 11 | Loss: 0.8183 | Time: 44.5s


Epoch 12 | Loss: 0.8150 | Time: 43.2s


Epoch 13 | Loss: 0.8122 | Time: 43.6s


Epoch 14 | Loss: 0.8116 | Time: 41.5s


Epoch 15 | Loss: 0.8099 | Time: 41.7s
  Run 2 Test Accuracy: 0.8862

  --- Run 3/3 ---


Epoch  1 | Loss: 1.1970 | Time: 40.5s


Epoch  2 | Loss: 0.9573 | Time: 39.7s


Epoch  3 | Loss: 0.8975 | Time: 40.3s


Epoch  4 | Loss: 0.8716 | Time: 39.8s


Epoch  5 | Loss: 0.8561 | Time: 39.8s


Epoch  6 | Loss: 0.8447 | Time: 40.3s


Epoch  7 | Loss: 0.8369 | Time: 40.0s


Epoch  8 | Loss: 0.8317 | Time: 39.7s


Epoch  9 | Loss: 0.8258 | Time: 44.8s


Epoch 10 | Loss: 0.8231 | Time: 48.7s


Epoch 11 | Loss: 0.8200 | Time: 49.5s


Epoch 12 | Loss: 0.8153 | Time: 49.3s


Epoch 13 | Loss: 0.8125 | Time: 49.7s


Epoch 14 | Loss: 0.8134 | Time: 46.1s


Epoch 15 | Loss: 0.8096 | Time: 46.7s
  Run 3 Test Accuracy: 0.8875


All runs complete!


## 10. Results Summary

In [10]:
print(f"{'Architecture':<15s} {'Mean Acc':>10s} {'Std Acc':>10s} {'Params':>12s} {'Time/Epoch':>12s}")
print('-' * 62)
for name, r in all_results.items():
    print(f"{name:<15s} {r['mean_acc']:>10.4f} {r['std_acc']:>10.4f} {r['params']:>12,} {r['mean_time_per_epoch']:>10.2f}s")

Architecture      Mean Acc    Std Acc       Params   Time/Epoch
--------------------------------------------------------------
1RNN                0.8687     0.0064    2,136,284       9.84s
1Bi-RNN             0.8675     0.0031    2,147,164      15.18s
2Bi-RNN             0.8533     0.0056    2,171,996      22.52s
1LSTM               0.8816     0.0023    2,168,156      14.60s
1Bi-LSTM            0.8804     0.0010    2,210,908      23.63s
2Bi-LSTM            0.8864     0.0008    2,310,236      43.20s


### Results Table colab

| | 1RNN | 1Bi-RNN | 2Bi-RNN | 1LSTM | 1Bi-LSTM | 2Bi-LSTM |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: |
| **Mean Accuracy** | 0.8709 | 0.8683 | 0.8585 | 0.8840 | 0.8832 | 0.8839 |
| **Std. Accuracy** | 0.0015 | 0.0004 | 0.0066 | 0.0037 | 0.0008 | 0.0037 |
| **Parameters** | 2,136,284 | 2,147,164 | 2,171,996 | 2,168,156 | 2,210,908 | 2,310,236 |
| **Time cost** | 4.96s | 5.16s | 5.35s | 5.32s | 5.64s | 5.76s |

### Results Table CPU 

| | 1RNN | 1Bi-RNN | 2Bi-RNN | 1LSTM | 1Bi-LSTM | 2Bi-LSTM |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **Mean Accuracy** | 0.8687 | 0.8675 | 0.8533 | 0.8816 | 0.8804 | 0.8864 |
| **Std. Accuracy** | 0.0064 | 0.0031 | 0.0056 | 0.0023 | 0.0010 | 0.0008 |
| **Parameters** | 2,136,284 | 2,147,164 | 2,171,996 | 2,168,156 | 2,210,908 | 2,310,236 |
| **Time cost** | 9.84s | 15.18s | 22.52s | 14.60s | 23.63s | 43.20s |

### **Your Observations:**

> **Q: How does model complexity (number of layers, bidirectionality, RNN vs LSTM) affect performance and training time?**

*[Your answer here]*

> **Q: Does adding a second layer help improve accuracy?**

*[Your answer here]*

> **Q: How stable are the models across the 3 runs (look at std accuracy)?**

*[Your answer here]*

## 11. t-SNE of Learned 1RNN Embeddings

In [11]:
tsne_words = [
    'business', 'career', 'student', 'university', 'college',
    'education', 'teacher', 'professor', 'school', 'degree',
    'economy', 'market', 'finance', 'investment', 'bank',
    'company', 'startup', 'entrepreneur', 'manager', 'salary',
    'science', 'research', 'technology', 'engineering', 'mathematics',
    'doctor', 'lawyer', 'accountant'
]

# Extract embeddings from the trained 1RNN model
rnn1_model = trained_models.get('1RNN')
if rnn1_model is None:
    print('ERROR: 1RNN model not found in trained_models dict.')
else:
    embedding_weights = rnn1_model.embedding_layer.weight.data.cpu().numpy()
    word_vectors = []
    valid_words = []
    for word in tsne_words:
        idx = vocab[word]  # returns <UNK> index if not found
        if idx != vocab['<UNK>']:
            word_vectors.append(embedding_weights[idx])
            valid_words.append(word)
        else:
            print(f"'{word}' not in vocabulary, skipping.")

    word_vectors = np.array(word_vectors)
    print(f'Collected {len(valid_words)} word vectors')

    # t-SNE
    tsne = TSNE(n_components=2, random_state=42, perplexity=8, n_iter=2000)
    embeddings_2d = tsne.fit_transform(word_vectors)

    plt.figure(figsize=(14, 10))
    plt.scatter(embeddings_2d[:, 0], embeddings_2d[:, 1],
                c='coral', s=100, alpha=0.7, edgecolors='darkred', linewidths=0.5)
    for i, word in enumerate(valid_words):
        plt.annotate(word, xy=(embeddings_2d[i, 0], embeddings_2d[i, 1]),
                     xytext=(7, 4), textcoords='offset points',
                     fontsize=11, fontweight='bold', color='darkslategray')

    plt.title('t-SNE of Learned 1RNN Embeddings (AG News)', fontsize=15, fontweight='bold')
    plt.xlabel('t-SNE Dim 1', fontsize=12)
    plt.ylabel('t-SNE Dim 2', fontsize=12)
    plt.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig('tsne_rnn_part_c.png', dpi=150, bbox_inches='tight')
    plt.show()
    print("Plot saved as 'tsne_rnn_part_c.png'")

Collected 28 word vectors


TypeError: TSNE.__init__() got an unexpected keyword argument 'n_iter'

### **Your Observations:**

> **Q: Compare the learned 1RNN t-SNE plot to the pre-trained GloVe plot from Part A. What differences do you notice? Why might they differ?**

*[Your answer here]*

## 12. Running Variations

To run each variation, go back to **Section 2 (Hyperparameters)** and change the toggle variables:

| Variation | Settings |
|-----------|----------|
| MAX_WORDS=50 | `MAX_WORDS = 50` |
| GloVe init (trainable) | `USE_GLOVE = True`, `FREEZE_EMBEDDINGS = False` |
| GloVe init (frozen) | `USE_GLOVE = True`, `FREEZE_EMBEDDINGS = True` |
| IMDB dataset | `USE_IMDB = True` |

Then **re-run all cells** from Section 3 onwards.

### **Your Observations on Variations:**

> **Q: How does increasing MAX_WORDS to 50 affect accuracy and training time?**

*[Your answer here]*

> **Q: Does pre-trained GloVe initialization help? What about freezing vs. fine-tuning?**

*[Your answer here]*

> **Q: How do results compare on IMDB (binary sentiment) vs AG News (4-class topic)?**

*[Your answer here]*